### PR TITLE
Add branch prediction for assert 

### PR DESCRIPTION
**Pull request by Gary-Hobson**
## Summary
Add a branches predicted compiler attribute

### CODE REVIEW

File: include/nuttx/compiler.h [hunk 1]
```diff
@@ -154,6 +154,11 @@
 #    define farcall_function __attribute__((long_call))
 #  endif
 
+/* Branch prediction */
+
+#define likely(x) __builtin_expect((x), 1)
+#define unlikely(x) __builtin_expect((x), 0)
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:160]: need add dummy definition for other compiler

Comment (Gary-Hobson) [reply include/nuttx/compiler.h]: done

Comment (pkarashchenko) [reply include/nuttx/compiler.h]: It will not work this way :D

Comment (xiaoxiang781216) [reply include/nuttx/compiler.h]: please follow how other macro is defined in this file, @Gary-Hobson

File: include/nuttx/compiler.h [hunk 2]
```diff
@@ -471,6 +476,8 @@
  */
 
 #  define noreturn_function
+#  define likely(x) x
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:479]: add () for all x

Comment (Gary-Hobson) [reply include/nuttx/compiler.h]: done

File: include/nuttx/compiler.h [hunk 3]
```diff
@@ -878,6 +891,8 @@
 #  define noreturn_function
 #  define farcall_function
 #  define aligned_data(n)
+#  define likely(x) (x)
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:894]: move before aligned_data

File: include/nuttx/compiler.h [hunk 4]
```diff
@@ -816,6 +827,8 @@
...
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:830]: move before aligned_data

File: include/nuttx/compiler.h [hunk 5]
```diff
@@ -622,6 +629,8 @@
 
 #  define noreturn_function
 #  define aligned_data(n)
+#  define likely(x) (x)
```
Comment (xiaoxiang781216) [inline include/nuttx/compiler.h:632]: move before aligned_data

**Review by xiaoxiang781216 — APPROVED** [review]

**Pull request merged**